In [44]:
import pandas as pd 


In [88]:
df_EPV=pd.read_excel(r"D:\Camara_Comercio\CCB_EPV_ANALISIS\db_EPV\Base EPV 2024 anonimizada.xlsx", sheet_name="ALFABETICO")

In [3]:
df_EPV.shape

(26064, 313)

In [4]:
df_EPV.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26064 entries, 0 to 26063
Columns: 313 entries, ID to FACTOR
dtypes: float64(39), int64(6), object(268)
memory usage: 62.2+ MB


In [6]:
df_EPV.head()

ID                START                  END  \
0   1  2024-06-04 14:40:25  2024-06-04 14:52:58   
1   3  2024-06-04 15:18:16  2024-06-04 15:25:31   
2   4  2024-06-04 15:24:59  2024-06-04 15:37:48   
3   6  2024-06-04 15:57:06  2024-06-04 16:04:38   
4   7  2024-06-04 15:50:58  2024-06-04 16:06:25   

  MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta LOCALIDAD  \
0                                             Bogotá                    Suba   
1                                             Bogotá                    Suba   
2                                             Bogotá                    Suba   
3                                             Bogotá                    Suba   
4                                             Bogotá                    Suba   

  SALUDO_Buenos días / tardes / noches. Mi nombre es _ y hago parte de GLORIA GALLLEGO SIGMA DOS INTERNACIONAL SAS, una firma de investigación y consultoría que puede consultar en la página web https://www.sigmados.com.co o en el teléfono 315 339 4797. Estamos haciendo un estudio para la Cámara de Comercio de Bogotá, sobre la percepción de las condiciones de seguridad en Bogotá, por lo que me gustaría hacerle unas preguntas. Su cooperación es voluntaria y muy importante para el estudio, su participación representa la de muchos ciudadanos. Sus respuestas serán estrictamente confidenciales y serán usadas solamente con fines estadísticos. Le tomará aproximadamente 20 minutos de su tiempo ¿Podría usted dedicarme unos minutos para responder algunas preguntas? Esta encuesta puede ser presenciada para asegurar la calidad de la actividad.  \
0                                                 Sí                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
1                                                 Sí                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
2                                                 Sí                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 

## Cantidad de nulos

In [10]:
df_EPV.shape

(26064, 313)

In [9]:
df_EPV.isnull().sum()

ID                                                                     0
START                                                                  0
END                                                                    0
MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta       0
LOCALIDAD                                                           6710
                                                                    ... 
DURACION_                                                              0
COD_DANE_SECTOR_                                                       0
COD_DANE_SECCION_                                                      0
COD_DANE_MANZANA_                                                      0
FACTOR                                                                 0
Length: 313, dtype: int64

# columnas con mayor cantidad de nulos 

## Actividades principales

-  renombrar columnas 
(MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta

SALUDO_Buenos días / tardes / noches. Mi nombre es _ y hago parte de GLORIA GALLLEGO SIGMA DOS INTERNACIONAL SAS, una firma de investigación y consultoría que puede consultar en la página web https://www.sigmados.com.co o en el teléfono 315 339 4797. Estamos haciendo un estudio para la Cámara de Comercio de Bogotá, sobre la percepción de las condiciones de seguridad en Bogotá, por lo que me gustaría hacerle unas preguntas. Su cooperación es voluntaria y muy importante para el estudio, su participación representa la de muchos ciudadanos. Sus respuestas serán estrictamente confidenciales y serán usadas solamente con fines estadísticos. Le tomará aproximadamente 20 minutos de su tiempo ¿Podría usted dedicarme unos minutos para responder algunas preguntas? Esta encuesta puede ser presenciada para asegurar la calidad de la actividad.

Sexo al nacer

Según factura de energía eléctrica estrato)



- estandarizar valores


- unir conjuntos de columnas

In [4]:
import pandas as pd

pd.set_option('display.max_columns', None)

In [5]:
df_EPV.columns

Index(['ID', 'START', 'END',
       'MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta',
       'LOCALIDAD',
       'SALUDO_Buenos días / tardes / noches. Mi nombre es _ y hago parte de GLORIA GALLLEGO SIGMA DOS INTERNACIONAL SAS, una firma de investigación y consultoría que puede consultar en la página web https://www.sigmados.com.co o en el teléfono 315 339 4797. Estamos haciendo un estudio para la Cámara de Comercio de Bogotá, sobre la percepción de las condiciones de seguridad en Bogotá, por lo que me gustaría hacerle unas preguntas. Su cooperación es voluntaria y muy importante para el estudio, su participación representa la de muchos ciudadanos. Sus respuestas serán estrictamente confidenciales y serán usadas solamente con fines estadísticos. Le tomará aproximadamente 20 minutos de su tiempo ¿Podría usted dedicarme unos minutos para responder algunas preguntas? Esta encuesta puede ser presenciada para asegurar la calidad de la actividad.',
       'Sexo al nacer', '

In [29]:
for col in df_EPV.columns:
    print(col)

ID
START
END
MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta
LOCALIDAD
SALUDO_Buenos días / tardes / noches. Mi nombre es _ y hago parte de GLORIA GALLLEGO SIGMA DOS INTERNACIONAL SAS, una firma de investigación y consultoría que puede consultar en la página web https://www.sigmados.com.co o en el teléfono 315 339 4797. Estamos haciendo un estudio para la Cámara de Comercio de Bogotá, sobre la percepción de las condiciones de seguridad en Bogotá, por lo que me gustaría hacerle unas preguntas. Su cooperación es voluntaria y muy importante para el estudio, su participación representa la de muchos ciudadanos. Sus respuestas serán estrictamente confidenciales y serán usadas solamente con fines estadísticos. Le tomará aproximadamente 20 minutos de su tiempo ¿Podría usted dedicarme unos minutos para responder algunas preguntas? Esta encuesta puede ser presenciada para asegurar la calidad de la actividad.
Sexo al nacer
Según factura de energía eléctrica estrato
Rango de edad


# diccionario

In [6]:
EPV2024_diccionario=pd.read_excel(r"D:\Camara_Comercio\Diccionario EPV 2024.xlsx")

In [7]:
EPV2024_diccionario

Id_Encuesta       Id_Pregunta  \
0    Encuesta de victimizacion y cultura ciudadana                ID   
1    Encuesta de victimizacion y cultura ciudadana             START   
2    Encuesta de victimizacion y cultura ciudadana               END   
3    Encuesta de victimizacion y cultura ciudadana         MUNICIPIO   
4    Encuesta de victimizacion y cultura ciudadana               NaN   
..                                             ...               ...   
700  Encuesta de victimizacion y cultura ciudadana          DURACION   
701  Encuesta de victimizacion y cultura ciudadana   COD_DANE_SECTOR   
702  Encuesta de victimizacion y cultura ciudadana  COD_DANE_SECCION   
703  Encuesta de victimizacion y cultura ciudadana  COD_DANE_MANZANA   
704  Encuesta de victimizacion y cultura ciudadana            FACTOR   

                                         Id_Pregunta.1  Id_Opcion  Opcion  
0                                                index        NaN     NaN  
1                                                start        NaN     NaN  
2                                                  end        NaN     NaN  
3    Seleccione el municipio donde va a aplicar la ...    11001.0  Bogotá  
4                                                  NaN    25126.0  Cajicá  
..                                                 ...        ...     ...  
700                                           Duracion        NaN     NaN  
701                                    COD_DANE_SECTOR        NaN     NaN  
702                                   COD_DANE_SECCION        NaN     NaN  
703                                   COD_DANE_MANZANA        NaN     NaN  
704                                Factor de expansion        NaN     NaN  

[705 rows x 5 columns]

In [16]:
EPV2024_diccionario.shape

(26064, 315)

# estandarizaciones de preguntas de escalas

### diccionario de escalas

In [68]:
import pandas as pd
import numpy as np

# Diccionario de equivalencias para escala 1–5
mapa_escala = {
    "Muy inseguro": 1,
    "Muy mala": 1,
    "Pésimo"
    "Inseguro": 2,
    "Ni seguro ni inseguro": 3,
    "Seguro": 4,
    "Muy seguro": 5,
    "No sabe /No opina":0,
    "Totalmente de acuerdo":1,
    "Totalmente en desacuerdo":2,
    
    # Por si vienen en minúscula o con variaciones
    "muy inseguro": 1,
    "inseguro": 2,
    "ni seguro ni inseguro": 3,
    "seguro": 4,
    "muy seguro": 5,
    "Muy buena": 5,
    "Excelente": 5,
    
    # asignacion de frecuencia de accion P407.1_¿Con qué frecuencia ve usted al policía del cuadrante?  
    "Alguna vez en el último semestre": 1,
    "De vez en cuando en el mes": 2,
    "Algunos días de la semana": 3, 
    "Todos los días": 4,

    # P106_Con respecto a Bogotá, en comparación con el año anterior ¿Usted cree que la inseguridad en lo corrido de este año 2024 …:  
    "Sigue igual":0,
    "Ha disminuido":1,
    "Ha aumentado": 2,

   # Sexo
   "Femenino": 0,
   "Masculino":1


}

columnas que se necesitan estandarizar

P308_En una escala 1 a 5, donde 1 es muy mala y 5 es muy buena, ¿Cómo califica la última atención que le brindaron en la llamada? 

In [ ]:
columnas_escala = [
    "P1101_En una escala de 1 a 5, donde 1 es muy inseguro y 5 es muy seguro. ¿Qué tan inseguros son cada uno de los siguientes espacios?",
    "P1102_Paraderos",
    "P1103_Parques",
    "P1104_Potreros",
    "P1105_Puentes peatonales",
    "P1106_Puentes vehiculares",
    "P1107_Semáforos",
    "P1108_Ciclo vías",
    "P1109_Ciclo rutas",
    "P111_En una escala de 1 a 5, donde 1 es muy inseguro y 5 es muy seguro. ¿Usted considera que el TransMilenio es….? ",
    "P446_En una escala de 1 a 5, donde 1 es muy inseguro y 5 es muy seguro. ¿Usted considera que, en términos de seguridad ciudadana, violencia y convivencia, Bogotá es seguro para hacer negocios?",
    "P1021_En una escala de 1 a 5, donde 1 es muy inseguro y 5 es muy seguro. El barrio en el que usted habita es…. ",
    "P1031_En una escala de 1 a 5, donde 1 es muy inseguro y 5 es muy seguro. Usted considera que Bogotá es…. ",
    "P449_¿Cómo calificaría su relacionamiento con las entidades de la Administración Distrital  ?",
    "P308_En una escala 1 a 5, donde 1 es muy mala y 5 es muy buena, ¿Cómo califica la última atención que le brindaron en la llamada? ",
    "P4011_¿Cómo califica el servicio prestado por la Policía? ", 

    # conjunto P4121_En la siguiente escala califique las siguientes habilidades del policía de su cuadrante en su última interacción:  1.Saludo amable
    "P4121_En la siguiente escala califique las siguientes habilidades del policía de su cuadrante en su última interacción:  1.Saludo amable",
    "P4122_2.Contacto visual",
    "P4123_3.Comunicación no verbal" ,
    "P4124_4.Amabilidad en el trato",   
    "P4125_5.Entendimiento de la necesidad", 
    "P4126_6. Disposición para ayudarlo",

    #pregunta individual

    "P444.1_En una escala de 1 a 5 donde 1 Muy mal y 5 es Muy Bien, cómo calificaría su conocimiento sobre el Código Nacional de Policía y Convivencia  ",
    "P421_En una escala de 1 a 5, donde 1 es pésimo y 5 es excelente, ¿Cómo se sintió con la atención prestada? ",
    "P113_¿Con qué frecuencia usa Transmilenio?",
    "P106_Con respecto a Bogotá, en comparación con el año anterior ¿Usted cree que la inseguridad en lo corrido de este año 2024 …:  ",
    
    #  "P4501_En una escala 1 a 5, donde 1 es Totalmente de acuerdo  y 5 Totalmente en desacuerdo, ¿Cómo califica el desempeño de la justicia en las siguientes acciones? 1) Los jueces liberan a personas que deberían estar recluidas en una cárcel o centro penitenciario."
    "P4501_En una escala 1 a 5, donde 1 es Totalmente de acuerdo  y 5 Totalmente en desacuerdo, ¿Cómo califica el desempeño de la justicia en las siguientes acciones? 1) Los jueces liberan a personas que deberían estar recluidas en una cárcel o centro penitenciario.",
    "P4502_2) Las penas que imponen los jueces son muy cortas",
    "P4503_3) Las penas que imponen los jueces son muy largas.",
    "P4504_4) En muchos casos la justicia no llega a condenar a las personas culpables.",
    "P4505_5) Los tiempos en que la justicia llega a decisiones finales son largos.",
    "P4506_6) Quienes pagan penas dentro de cárceles o centros penitenciarios pueden resocializarse.",

    #pregunta individual
    "P407.1_¿Con qué frecuencia ve usted al policía del cuadrante?  ",

    "Sexo al nacer"

]

In [87]:
for col in columnas_escala:
    df_EPV[col] = (
        df_EPV[col]
        .astype(str)                      # convertir todo a string
        .str.strip()                      # quitar espacios
        .replace(mapa_escala)             # reemplazar texto por número
    )
    
    # Convertir a numérico
    df_EPV[col] = pd.to_numeric(df_EPV[col], errors="coerce")

C:\Users\MI EQUIPO\AppData\Local\Temp\ipykernel_19556\2286002656.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .replace(mapa_escala)             # reemplazar texto por número
C:\Users\MI EQUIPO\AppData\Local\Temp\ipykernel_19556\2286002656.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .replace(mapa_escala)             # reemplazar texto por número
C:\Users\MI EQUIPO\AppData\Local\Temp\ipykernel_19556\2286002656.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version.

# estandarización de preguntas con si/no

In [71]:
columnas_binarias = [
    "P230_¿algún miembro de su hogar (excluyéndose a usted mismo) fue víctima de algún delito? HOGAR Es la persona o grupo de personas, parientes o no, que ocupan la totalidad o parte de una vivienda; atienden necesidades básicas, con cargo a un presupuesto común y generalmente comparten las comidas. DANE. ",
    "P233_En el  2024, ¿Usted afrontó alguna situación que afectó su convivencia? (Como riñas, reuniones ruidosas de noche, arrojar basura a un sitio público, entre otros)  ",
    "P305_En este año 2024, ¿Usted ha llamado al número de atención de emergencia 123?  ",
    "P3051_¿Su llamada al 123, estuvo relacionada con denunciar algún hecho/delito de seguridad o convivencia?",
    # grupo preguntas P4511
    "P4511_En los últimos 12 meses ¿Ha participado en alguno de los siguientes espacios comunitarios para discutir temas de Seguridad y Convivencia? ",
    "P4512_2. Frentes Locales de Seguridad",
    "P4513_3. Juntas Zonales de Seguridad",
    "P4514_4. Sesiones de la Junta Administradora local donde traten temas de seguridad",
    "P4515_5. Reuniones con vecinos con participación de funcionarios de alcaldía local o distrital",
    # hay muy pocos que contestaron "si" en otro
    "P4516_6. Otro",
    "P4517_Cuál ? Normalizada",
    # grupo preguntas P4521
    "P4521_¿En cuáles de las siguientes situaciones justificaría el usar la violencia? 1. Cuando es la única manera de alcanzar un objetivo",
    "P4522_2. Cuando es la única forma de ayudarle a un familiar o amigo",
    "P4523_3. Cuando se lucha contra una ley o régimen injusto",
    "P4524_4. Por provecho económico",
    "P4525_5. Cuando la creencia religiosa lo permite o exige",
    "P4526_6. Para responder ante una ofensa",
    "P4527_7. Cuando se tiene certeza de que no habrá un castigo",
    "P4528_8. Cuando otras personas lo hacen y les ha ido bien.",
    "P4529_9. Cuando es lo acostumbrado", 
    "P45210_10. Para defender las propiedades o bienes propios",
    "P45211_11. Para pagar un favor.", 
    #pregunta individual
    "P405_¿Usted sabe a cuál cuadrante pertenece su hogar?  ",
    # grupo preguntas P4071
    "P4071_En este 2024, en el cuadrante donde usted vive… 1.Ha visto al policía del cuadrante",
    "P4072_2.Ha interactuado con el policía de su cuadrante",
    "P4073_3. Usted conoce el número de celular de la Policía del cuadrante",
    #pregunta individual
    "P417_En este 2024, ¿Usted acudió a la Policía?  ",
    "P444_¿Usted conoce el Código Nacional de Policía y Convivencia? ",
    "P500_¿Usted tiene un negocio, empresa o establecimiento de comercio?",
    "P501_¿Este negocio, empresa o establecimiento de comercio tiene matricula mercantil?.",
    "P102_El barrio en el que usted habita es…. ", 
    "P103_Usted considera que Bogotá es…………."
    

]

In [72]:
mapa_binario = {
    "Sí": 1,
    "Si": 1,
    "sí": 1,
    "si": 1,
    "No": 0,
    "no": 0,
    "No sabe": 2,
    "NS/NR": 2,
    "No responde": 2,
    # P102_El barrio en el que usted habita es…. 
    # P103_Usted considera que Bogotá es………….
    "Inseguro/ a":0,
    "Seguro/a":1
}

In [73]:
for col in columnas_binarias:
    df_EPV[col] = (
        df_EPV[col]
        .astype(str)
        .str.strip()
        .replace(mapa_binario)
    )
    
    df_EPV[col] = pd.to_numeric(df_EPV[col], errors="coerce")

C:\Users\MI EQUIPO\AppData\Local\Temp\ipykernel_19556\850882666.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .replace(mapa_binario)
C:\Users\MI EQUIPO\AppData\Local\Temp\ipykernel_19556\850882666.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .replace(mapa_binario)
C:\Users\MI EQUIPO\AppData\Local\Temp\ipykernel_19556\850882666.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To o

In [74]:
df_EPV[columnas_binarias]

P230_¿algún miembro de su hogar (excluyéndose a usted mismo) fue víctima de algún delito? HOGAR Es la persona o grupo de personas, parientes o no, que ocupan la totalidad o parte de una vivienda; atienden necesidades básicas, con cargo a un presupuesto común y generalmente comparten las comidas. DANE.   \
0                                                      0                                                                                                                                                                                                                                                                 
1                                                      2                                                                                                                                                                                                                                                                 
2                                                      0                                                                                                                                                                                                                                                                 
3                                                      0                                                                                                                                                                                                                                                                 
4                                                      0                                                                                                                                                                                                                                                                 
...                                                  ...                                                                                                                                                                                                                                                                 
26059                                                  0                                                                                                                                                                                                                                                                 
26060                                                  0                                                                                                                                                                                                                                                                 
26061                                                  0                                                                                                                                                                                                                                                                 
26062                                                  2                                                                                                                                                                                                                                                                 
26063                                                  0                                                                                                                                                                                                                                                                 

       P233_En el  2024, ¿Usted afrontó alguna situación que afectó su convivencia? (Como riñas, reuniones ruidosas de noche, arrojar basura a un sitio público, entre otros)    \
0                                                      0   

In [75]:
df_EPV[columnas_binarias].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26064 entries, 0 to 26063
Data columns (total 32 columns):
 #   Column                                                                                                                                                                                                                                                                                                           Non-Null Count  Dtype  
---  ------                                                                                                                                                                                                                                                                                                           --------------  -----  
 0   P230_¿algún miembro de su hogar (excluyéndose a usted mismo) fue víctima de algún delito? HOGAR Es la persona o grupo de personas, parientes o no, que ocupan la totalidad o parte de una vivienda; atienden necesidades b

# ******************  unficiacion multiples respuestas ***************

In [14]:
import numpy as np
import pandas as pd

# Diccionario: columna -> categoría general
mapa_categorias = {
    "P1121_¿En cuanto a los siguientes aspectos, ¿Cuáles son las tres principales razones por la que considera a TransMilenio inseguro o muy inseguro?  /1. Estaciones, portales, paraderos o buses muy llenos": "Congestión / Operación",
    
    "P1122_Presencia de habitantes de calle": "Desorden social",
    "P1123_Presencia de vendedores ambulantes, cantantes, personas pidiendo dinero, etc.": "Desorden social",
    "P1126_Colados": "Desorden social",
    "P1128_Protestas, bloqueos, vandalismo": "Desorden social",
    
    "P1124_Actividades ilegales alrededor de las estaciones": "Delincuencia directa",
    "P1127_Peleas, riñas o intolerancias al interior de la estación y los buses": "Delincuencia directa",
    "P11211_Hurtos (atraco, cosquilleo, raponazo)": "Delincuencia directa",
    "P11212_Acoso sexual, manoseos , piropos, miradas o actos obscenos": "Delincuencia directa",
    
    "P1125_Espacios abandonados o deteriorados cerca a las estaciones.": "Infraestructura deteriorada",
    "P11210_Estaciones deterioradas (dañadas, vandalizadas, con poca iluminación o sin puertas)": "Infraestructura deteriorada",
    
    "P1129_Falta de presencia policial o de personal de seguridad": "Falta de control institucional"
}

In [15]:
def agrupar_respuestas(row):
    categorias = set()
    
    for col, categoria in mapa_categorias.items():
        if pd.notna(row[col]):  # si la persona marcó esa opción
            categorias.add(categoria)
    
    if len(categorias) == 0:
        return np.nan
    
    return ", ".join(categorias)


df_EPV["P1121_categoria_general"] = df_EPV.apply(agrupar_respuestas, axis=1)

In [16]:
mapa_codigos = {
    "Delincuencia directa": 1,
    "Desorden social": 2,
    "Infraestructura deteriorada": 3,
    "Falta de control institucional": 4,
    "Congestión / Operación": 5
}

df_EPV["P1121_¿En cuanto a los siguientes aspectos, ¿Cuáles son las tres principales razones por la que considera a TransMilenio inseguro o muy inseguro?"] = df_EPV["P1121_categoria_general"].map(mapa_codigos)

In [17]:
df_EPV.drop(columns=mapa_categorias.keys(), inplace=True)

In [18]:
df_EPV["P1121_¿En cuanto a los siguientes aspectos, ¿Cuáles son las tres principales razones por la que considera a TransMilenio inseguro o muy inseguro?"]

0        NaN
1        2.0
2        5.0
3        NaN
4        NaN
        ... 
26059    NaN
26060    NaN
26061    NaN
26062    NaN
26063    1.0
Name: P1121_¿En cuanto a los siguientes aspectos, ¿Cuáles son las tres principales razones por la que considera a TransMilenio inseguro o muy inseguro?, Length: 26064, dtype: float64

In [19]:
df_EPV["P1121_¿En cuanto a los siguientes aspectos, ¿Cuáles son las tres principales razones por la que considera a TransMilenio inseguro o muy inseguro?"].value_counts()

P1121_¿En cuanto a los siguientes aspectos, ¿Cuáles son las tres principales razones por la que considera a TransMilenio inseguro o muy inseguro?
1.0    806
2.0    663
5.0    503
4.0    244
3.0     28
Name: count, dtype: int64

# dejar las columnas como one hot ()


## preguntas que tienen como valor el nombre de la columna

In [76]:
# P1121_¿En cuanto a los siguientes aspectos, ¿Cuáles son las tres principales razones por la que considera a TransMilenio inseguro o muy inseguro?  /1. Estaciones, portales, paraderos o buses muy llenos
df_EPV.loc[:, "P1121_¿En cuanto a los siguientes aspectos, ¿Cuáles son las tres principales razones por la que considera a TransMilenio inseguro o muy inseguro?  /1. Estaciones, portales, paraderos o buses muy llenos"
              :"P11212_Acoso sexual, manoseos , piropos, miradas o actos obscenos"] \
      = df_EPV.loc[:, "P1121_¿En cuanto a los siguientes aspectos, ¿Cuáles son las tres principales razones por la que considera a TransMilenio inseguro o muy inseguro?  /1. Estaciones, portales, paraderos o buses muy llenos"
              :"P11212_Acoso sexual, manoseos , piropos, miradas o actos obscenos"] \
        .notna().astype(int)
 #P1171_De las siguientes modalidades delictivas, ¿Cuáles son las tres que más le preocupa?
df_EPV.loc[:, "P1171_De las siguientes modalidades delictivas, ¿Cuáles son las tres que más le preocupa?"
              :"P11721_Abigeato"] \
      = df_EPV.loc[:, "P1171_De las siguientes modalidades delictivas, ¿Cuáles son las tres que más le preocupa?"
              :"P11721_Abigeato"] \
        .notna().astype(int)

#P1281_Con relación a la seguridad de Bogotá ¿Cuáles diría usted que son las tres principales fuentes de información que le permite formar una opinión sobre los temas de seguridad en el municipio?  
df_EPV.loc[:, "P1281_Con relación a la seguridad de Bogotá ¿Cuáles diría usted que son las tres principales fuentes de información que le permite formar una opinión sobre los temas de seguridad en el municipio?  "
              :"P1289_Páginas oficiales en internet (Alcaldía, Fiscalía, Policía, Medicina Legal)."] \
      = df_EPV.loc[:, "P1281_Con relación a la seguridad de Bogotá ¿Cuáles diría usted que son las tres principales fuentes de información que le permite formar una opinión sobre los temas de seguridad en el municipio?  "
              :"P1289_Páginas oficiales en internet (Alcaldía, Fiscalía, Policía, Medicina Legal)."] \
        .notna().astype(int)

# P621_¿En qué actividad ocupa la mayor parte del tiempo?  

df_EPV.loc[:, "P621_¿En qué actividad ocupa la mayor parte del tiempo?  "
              :"P6210_No responde"] \
      = df_EPV.loc[:, "P621_¿En qué actividad ocupa la mayor parte del tiempo?  "
              :"P6210_No responde"] \
        .notna().astype(int)



## preguntas con valores si/no

In [77]:
mapa_binario = {
    "Sí": 1,
    "Si": 1,
    "sí": 1,
    "si": 1,
    "No": 0,
    "no": 0,
    "No sabe": 2,
    "NS/NR": 2,
    "No responde": 2,
    # P102_El barrio en el que usted habita es…. 
    # P103_Usted considera que Bogotá es………….
    "Inseguro/ a":0,
    "Seguro/a":1
}

P2041_De la siguiente lista indíqueme de qué delito o delitos fue víctima Hurto a personas (Atraco, cosquilleo, raponazo, fraude, engaño, paseo millonario, etc)


P2042211_El medio utilizado en la extorsión fueDirecto

In [ ]:

#P204111_¿En cuanto al hurto a persona, bajo qué modalidad se cometió el delito o delitos? Fleteo
df_EPV.loc[:, "P204111_¿En cuanto al hurto a persona, bajo qué modalidad se cometió el delito o delitos? Fleteo"
              :"P204119_Cuál ? Normalizado"] = \
df_EPV.loc[:, "P204111_¿En cuanto al hurto a persona, bajo qué modalidad se cometió el delito o delitos? Fleteo"
              :"P204119_Cuál ? Normalizado"] \
    .replace({
        "Sí": 1, "Si": 1, "sí": 1, "si": 1,
        "No": 0, "no": 0
    })

#P2081_¿Qué le hurtaron?   Billetera

df_EPV.loc[:, "P2081_¿Qué le hurtaron?   Billetera"
              :"P2087_Otro "] = \
df_EPV.loc[:, "P2081_¿Qué le hurtaron?   Billetera"
              :"P2087_Otro "] \
    .replace({
        "Sí": 1, "Si": 1, "sí": 1, "si": 1,
        "No": 0, "no": 0
    })


# P204121_¿En cuanto a los Delitos Cibernéticos del que fue víctima, bajo qué modalidad se cometió?  Recepción de mensajes de texto (SMS), simulando ser parte de una entidad para hurtar información privada.

df_EPV.loc[:, "P204121_¿En cuanto a los Delitos Cibernéticos del que fue víctima, bajo qué modalidad se cometió?  Recepción de mensajes de texto (SMS), simulando ser parte de una entidad para hurtar información privada."
              :"P204124_Secuestro de datos"] = \
df_EPV.loc[:, "P204121_¿En cuanto a los Delitos Cibernéticos del que fue víctima, bajo qué modalidad se cometió?  Recepción de mensajes de texto (SMS), simulando ser parte de una entidad para hurtar información privada."
              :"P204124_Secuestro de datos"] \
    .replace({
        "Sí": 1, "Si": 1, "sí": 1, "si": 1,
        "No": 0, "no": 0
    })


#P2341_¿Qué situación o situaciones que afectan la convivencia afrontó este año?  1. Arrojar basuras en sitio público

df_EPV.loc[:, "P2341_¿Qué situación o situaciones que afectan la convivencia afrontó este año?  1. Arrojar basuras en sitio público"
              :"P23430_Invasión del espacio público"] = \
df_EPV.loc[:, "P2341_¿Qué situación o situaciones que afectan la convivencia afrontó este año?  1. Arrojar basuras en sitio público"
              :"P23430_Invasión del espacio público"] \
    .replace({
        "Sí": 1, "Si": 1, "sí": 1, "si": 1,
        "No": 0, "no": 0
    })

# P4071_En este 2024, en el cuadrante donde usted vive… 1.Ha visto al policía del cuadrante

df_EPV.loc[:, "P4071_En este 2024, en el cuadrante donde usted vive… 1.Ha visto al policía del cuadrante"
              :"P4073_3. Usted conoce el número de celular de la Policía del cuadrante"] = \
df_EPV.loc[:, "P4071_En este 2024, en el cuadrante donde usted vive… 1.Ha visto al policía del cuadrante"
              :"P4073_3. Usted conoce el número de celular de la Policía del cuadrante"] \
    .replace({
        "Sí": 1, "Si": 1, "sí": 1, "si": 1,
        "No": 0, "no": 0
    })

# P2141_En este año 2024, ¿Usted denunció alguno de estos delitos? 

df_EPV["P2141_En este año 2024, ¿Usted denunció alguno de estos delitos? "].replace({"No":0, "Si":1})


C:\Users\MI EQUIPO\AppData\Local\Temp\ipykernel_19556\4198958665.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .replace({
C:\Users\MI EQUIPO\AppData\Local\Temp\ipykernel_19556\4198958665.py:17: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .replace({
C:\Users\MI EQUIPO\AppData\Local\Temp\ipykernel_19556\4198958665.py:29: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future

# preguntas de municipios y localidad

In [79]:
import pandas as pd
import numpy as np

# Seleccionar rango por nombre
municipios_cols = df_EPV.loc[:, 
    "P203231_¿En qué  municipios fue víctima de algún delito? Bogota"
    :"P2032360_Zipaquirá"
].columns

# Crear nueva columna concatenando respuestas no nulas
df_EPV["municipio_victima"] = (
    df_EPV[municipios_cols]
        .apply(lambda row: ", ".join(row.dropna().astype(str)), axis=1)
)

# Reemplazar vacío por NaN
df_EPV["P203231_¿En qué  municipios fue víctima de algún delito?"] = df_EPV["municipio_victima"].replace("", np.nan)

In [80]:
localidades_cols = df_EPV.loc[:,
    "P203241_¿En qué localidades de Bogotá fue víctima de algún delito?  Usaquen"
    :"P2032419_Ciudad Bolívar"
].columns

df_EPV["localidad_victima"] = (
    df_EPV[localidades_cols]
        .apply(lambda row: ", ".join(row.dropna().astype(str)), axis=1)
)

df_EPV["P203241_¿En qué localidades de Bogotá fue víctima de algún delito?"] = df_EPV["localidad_victima"].replace("", np.nan)

In [81]:
df_EPV["P203231_¿En qué  municipios fue víctima de algún delito? Bogota"]

0           NaN
1           NaN
2        Bogotá
3           NaN
4        Bogotá
          ...  
26059       NaN
26060       NaN
26061       NaN
26062       NaN
26063       NaN
Name: P203231_¿En qué  municipios fue víctima de algún delito? Bogota, Length: 26064, dtype: object

In [82]:
municipios_cols = df_EPV.loc[:, 
    "P203231_¿En qué  municipios fue víctima de algún delito? Bogota"
    :"P2032360_Zipaquirá"
].columns

df_EPV.drop(columns=municipios_cols, inplace=True)

In [83]:
localidades_cols = df_EPV.loc[:, 
    "P203241_¿En qué localidades de Bogotá fue víctima de algún delito?  Usaquen"
    :"P2032419_Ciudad Bolívar"
].columns

df_EPV.drop(columns=localidades_cols, inplace=True)

# guardar archivo

In [85]:
df_EPV.to_excel("DF_EPV_version2.xlsx", index=False)

In [90]:
df_EPV.shape

(26064, 313)

# Unificacion de columnas ( no ejecutar )

In [35]:
df = df_EPV.copy()
id_col = "ID"
peso = "FACTOR"

In [ ]:
bloques.append(procesar_multirespuesta(
    df,
    "P2042221_P204.22.2. Con respecto a la extorsión, usted:  Denunció ante las autoridades",
    "P2042224_No hizo nada y no volvieron a comunicarse",
    "Extorsión"
))

# 2 Modalidad hurto persona
bloques.append(procesar_multirespuesta(
    df,
    "P204111_¿En cuanto al hurto a persona, bajo qué modalidad se cometió el delito o delitos? Fleteo",
    "P204118_Otro ¿Cuál?",
    "Modalidad hurto persona"
))

# 4 Qué le hurtaron
bloques.append(procesar_multirespuesta(
    df,
    "P2081_¿Qué le hurtaron?   Billetera",
    "P2087_Otro ",
    "Qué le hurtaron"
))

# 5 Delitos cibernéticos
bloques.append(procesar_multirespuesta(
    df,
    "P204121_¿En cuanto a los Delitos Cibernéticos del que fue víctima, bajo qué modalidad se cometió?  Recepción de mensajes de texto (SMS), simulando ser parte de una entidad para hurtar información privada.",
    "P204124_Secuestro de datos",
    "Delitos cibernéticos"
))

# 7 Participación comunitaria
bloques.append(procesar_multirespuesta(
    df,
    "P4511_En los últimos 12 meses ¿Ha participado en alguno de los siguientes espacios comunitarios para discutir temas de Seguridad y Convivencia? ",
    "P4517_Cuál ? Normalizada",
    "Participación comunitaria"
))

# 8 Justificación violencia
bloques.append(procesar_multirespuesta(
    df,
    "P4521_¿En cuáles de las siguientes situaciones justificaría el usar la violencia? 1. Cuando es la única manera de alcanzar un objetivo",
    "P45211_11. Para pagar un favor.",
    "Justificación violencia"
))

# 9 Cuadrante 2024
bloques.append(procesar_multirespuesta(
    df,
    "P4071_En este 2024, en el cuadrante donde usted vive… 1.Ha visto al policía del cuadrante",
    "P4073_3. Usted conoce el número de celular de la Policía del cuadrante",
    "Policía cuadrante"
))

# 6 Escala justicia
bloques.append(procesar_escala(
    df,
    "P4501_En una escala 1 a 5, donde 1 es Totalmente de acuerdo  y 5 Totalmente en desacuerdo, ¿Cómo califica el desempeño de la justicia en las siguientes acciones? 1) Los jueces liberan a personas que deberían estar recluidas en una cárcel o centro penitenciario.",
    "P4506_6) Quienes pagan penas dentro de cárceles o centros penitenciarios pueden resocializarse.",
    "Percepción justicia"
))

# 10 Escala habilidades policía
bloques.append(procesar_escala(
    df,
    "P4121_En la siguiente escala califique las siguientes habilidades del policía de su cuadrante en su última interacción:  1.Saludo amable",
    "P4126_6. Disposición para ayudarlo",
    "Habilidades policía"
))

In [57]:
import pandas as pd

df = df_EPV.copy()

id_col = "ID"
peso = "FACTOR"

def convertir_a_columna_unica(df, col_inicio, col_fin):
    
    idx_inicio = df.columns.get_loc(col_inicio)
    idx_fin = df.columns.get_loc(col_fin)
    cols = df.columns[idx_inicio:idx_fin+1]
    
    # Extraer encabezado común (texto entre "_" y el número)
    primer_nombre = cols[0]
    encabezado = primer_nombre.split("_", 1)[1]
    encabezado = encabezado.rsplit("  ", 1)[0]  # quitar "1.Saludo amable"
    
    temp = df[[id_col, peso] + list(cols)]
    
    df_long = temp.melt(
        id_vars=[id_col, peso],
        value_vars=cols,
        var_name="col_original",
        value_name="valor"
    )
    
    # Solo donde marcaron (Si es tipo Sí/No)
    df_long = df_long[df_long["valor"] == "Sí"]
    
    # Extraer solo la opción (1.Saludo amable)
    df_long["opcion"] = df_long["col_original"].str.split(":").str[-1].str.strip()
    
    # Renombrar la columna con el encabezado real
    df_long = df_long[[id_col, "opcion", peso]]
    df_long = df_long.rename(columns={"opcion": encabezado})
    
    return df_long

In [59]:
df_habilidades = convertir_a_columna_unica(
    df,
    "P4121_En la siguiente escala califique las siguientes habilidades del policía de su cuadrante en su última interacción:  1.Saludo amable",
    "P4126_6. Disposición para ayudarlo",
    
)

In [60]:
df_habilidades

Empty DataFrame
Columns: [ID, En la siguiente escala califique las siguientes habilidades del policía de su cuadrante en su última interacción:, FACTOR]
Index: []

# unificacion version 2 (no ejecutar)

In [62]:
df_EPV.columns

Index(['ID', 'START', 'END',
       'MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta',
       'LOCALIDAD',
       'SALUDO_Buenos días / tardes / noches. Mi nombre es _ y hago parte de GLORIA GALLLEGO SIGMA DOS INTERNACIONAL SAS, una firma de investigación y consultoría que puede consultar en la página web https://www.sigmados.com.co o en el teléfono 315 339 4797. Estamos haciendo un estudio para la Cámara de Comercio de Bogotá, sobre la percepción de las condiciones de seguridad en Bogotá, por lo que me gustaría hacerle unas preguntas. Su cooperación es voluntaria y muy importante para el estudio, su participación representa la de muchos ciudadanos. Sus respuestas serán estrictamente confidenciales y serán usadas solamente con fines estadísticos. Le tomará aproximadamente 20 minutos de su tiempo ¿Podría usted dedicarme unos minutos para responder algunas preguntas? Esta encuesta puede ser presenciada para asegurar la calidad de la actividad.',
       'Sexo al nacer', '

In [ ]:
import pandas as pd
import numpy as np

df= df_EPV.copy()

# ==========================================================
# 1️⃣ DEFINIR LOS RANGOS
# ==========================================================

rangos = [
    ("P2042221_", "P2042224_"),
    ("P204111_", "P204118_"),
    ("P204118_", "P204119_"),
    ("P2081_", "P2087_"),
    ("P204121_", "P204124_"),
    ("P4501_", "P4506_"),
    ("P4511_", "P4517_"),
    ("P4521_", "P45211_"),
    ("P4071_", "P4073_"),
    ("P4121_", "P4126_"),
]

# ==========================================================
# 2️⃣ FUNCIÓN PARA OBTENER COLUMNAS POR RANGO
# ==========================================================

def obtener_columnas_por_rango(df, inicio, fin):
    cols = df.columns.tolist()
    idx_inicio = next(i for i, c in enumerate(cols) if c.startswith(inicio))
    idx_fin = next(i for i, c in enumerate(cols) if c.startswith(fin))
    return cols[idx_inicio:idx_fin+1]


# ==========================================================
# 3️⃣ OPCIÓN 1: UNA SOLA RESPUESTA (PRIMERA NO NULA)
# ==========================================================

df_opcion1 = df.copy()

for inicio, fin in rangos:
    cols_rango = obtener_columnas_por_rango(df, inicio, fin)
    
    nueva_col = cols_rango[0].split("_",1)[1]  # texto pregunta
    
    df_opcion1[nueva_col] = df_opcion1[cols_rango] \
        .apply(lambda row: next((col.split("_",1)[1] 
                                 for col, val in row.items() 
                                 if pd.notnull(val) and val not in ["No", 0]), 
                                np.nan), axis=1)
    
    df_opcion1.drop(columns=cols_rango, inplace=True)


# ==========================================================
# 4️⃣ OPCIÓN 2: CONCATENAR RESPUESTAS
# ==========================================================

df_opcion2 = df.copy()

for inicio, fin in rangos:
    cols_rango = obtener_columnas_por_rango(df, inicio, fin)
    
    nueva_col = cols_rango[0].split("_",1)[1]
    
    def concatenar(row):
        respuestas = []
        for col, val in row.items():
            if pd.notnull(val) and val not in ["No", 0]:
                respuestas.append(col.split("_",1)[1])
        if len(respuestas) == 0:
            return np.nan
        return ", ".join(respuestas)
    
    df_opcion2[nueva_col] = df_opcion2[cols_rango].apply(concatenar, axis=1)
    
    df_opcion2.drop(columns=cols_rango, inplace=True)


# ==========================================================
# 5️⃣ OPCIÓN 3: FORMATO LONG (REPETIR REGISTROS)
# ==========================================================

dfs_long = []

for inicio, fin in rangos:
    cols_rango = obtener_columnas_por_rango(df, inicio, fin)
    
    df_melt = df.melt(
        id_vars=[col for col in df.columns if col not in cols_rango],
        value_vars=cols_rango,
        var_name="Pregunta_Opcion",
        value_name="Valor"
    )
    
    df_melt = df_melt[
        df_melt["Valor"].notna() & 
        (df_melt["Valor"] != "No") & 
        (df_melt["Valor"] != 0)
    ]
    
    df_melt["Pregunta"] = cols_rango[0].split("_",1)[1]
    df_melt["Opcion"] = df_melt["Pregunta_Opcion"].apply(lambda x: x.split("_",1)[1])
    
    dfs_long.append(df_melt)

df_long = pd.concat(dfs_long, ignore_index=True)

# ==========================================================
# RESULTADOS
# ==========================================================

print("Tabla opción 1 lista: df_opcion1")
print("Tabla opción 2 lista: df_opcion2")
print("Tabla long lista: df_long")

KeyError: "['P204118_Otro ¿Cuál?'] not in index"

# version 3 (no ejecutar)

In [14]:
import pandas as pd

# 1. Definimos las columnas que NO queremos mover (las que se quedan fijas)
columnas_identificadores = [
    'ID', 'START', 'END', 'MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta', 
    'LOCALIDAD', 'Sexo al nacer', 'Según factura de energía eléctrica estrato', 
    'Rango de edad', 'Edad exacta'
]

# 2. Definimos las columnas de respuestas que queremos unificar (las que se convertirán en filas)
columnas_respuestas = [
    'P621_¿En qué actividad ocupa la mayor parte del tiempo?  ',
    'P622_Desempleado', 'P623_Estudiando', 'P624_Actividades de ocio y/o recreación',
    'P625_Oficios del Hogar', 'P626_Buscando trabajo', 
    'P627_Incapacitado Permanente para Trabajar', 'P628_Otro', 
    'P629_No sabe', 'P6210_No responde', 'P6211_Cual_Normalizada'
]

# 3. Transformamos la tabla de formato ancho a largo
df_largo = pd.melt(
    df_EPV, 
    id_vars=columnas_identificadores, 
    value_vars=columnas_respuestas,
    var_name='Pregunta_Original', 
    value_name='Respuesta'
)

# 4. Filtramos los nulos
# Como quieres que solo aparezcan las opciones que la gente marcó, eliminamos donde la respuesta sea nula
df_largo = df_largo.dropna(subset=['Respuesta'])

# 5. Opcional: Limpiar el nombre de la actividad para que no diga "P622_Desempleado" sino solo "Desempleado"
df_largo['Actividad_Principal'] = df_largo['Pregunta_Original'].str.split('_').str[-1]

# Mostramos el resultado
print(df_largo[['ID', 'Actividad_Principal']].sort_values(by='ID'))

           ID                                Actividad_Principal
104256      1                                  Oficios del Hogar
52128       1                                         Estudiando
52129       3                                         Estudiando
23094       4  ¿En qué actividad ocupa la mayor parte del tie...
130322      4                                   Buscando trabajo
...       ...                                                ...
19349   26713  ¿En qué actividad ocupa la mayor parte del tie...
19350   26714  ¿En qué actividad ocupa la mayor parte del tie...
19351   26715  ¿En qué actividad ocupa la mayor parte del tie...
19352   26716  ¿En qué actividad ocupa la mayor parte del tie...
19353   26717  ¿En qué actividad ocupa la mayor parte del tie...

[28796 rows x 2 columns]


In [15]:
df_largo

ID                       START                         END  \
4          7         2024-06-04 15:50:58         2024-06-04 16:06:25   
6         17  2024-06-04 09:34:40.591000  2024-06-04 09:53:10.978000   
9         20  2024-06-04 09:39:15.596000  2024-06-04 09:56:31.344000   
10        21  2024-06-04 09:00:23.031000  2024-06-04 09:55:27.190000   
11        23  2024-06-04 09:38:46.226000  2024-06-04 10:01:58.114000   
...      ...                         ...                         ...   
286589  4694  2024-06-29 12:29:31.800000  2024-06-29 12:42:03.389000   
286591  4696  2024-06-29 14:18:26.828000  2024-06-29 14:30:24.871000   
286602  4707  2024-06-29 12:56:05.733000  2024-06-29 13:09:56.700000   
286662  8066  2024-07-10 14:02:21.603000  2024-07-10 14:17:59.580000   
286685  8089  2024-07-10 14:13:35.702000  2024-07-10 14:34:02.113000   

       MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta  \
4                                                  Bogotá                 
6                                                  Bogotá                 
9                                                  Bogotá                 
10                                                 Bogotá                 
11                                                 Bogotá                 
...                                                   ...                 
286589                                          Zipaquirá                 
286591                                          Zipaquirá                 
286602                                          Zipaquirá                 
286662                                          Zipaquirá                 
286685                                          Zipaquirá                 

          LOCALIDAD Sexo al nacer  Según factura de energía eléctrica estrato  \
4              Suba      Femenino                                           2   
6              Usme     Masculino                                           2   
9              Suba      Femenino                                           2   
10      Teusaquillo     Masculino                                           4   
11         Engativá     Masculino                                           3   
...             ...           ...                                         ...   
286589          NaN     Masculino                                           4   
286591          NaN     Masculino                                           3   
286602          NaN     Masculino                                           2   
286662          NaN     Masculino                                           2   
286685          NaN      Femenino                                           3   

          Rango de edad  Edad exacta  \
4       De 66 a 75 años           66   
6       De 46 a 55 años           53   
9       De 56 a 65 años           65   
10      De 66 a 75 años           66   
11      De 56 a 65 años           60   
...                 ...          ...   
286589   Mas de 75 años           85   
286591   Mas de 75 años           92   
286602  De 66 a 75 años           68   
286662  De 66 a 75 años           66   
286685  De 56 a 65 años           65   

                                        Pregunta_Original  \
4       P621_¿En qué actividad ocupa la mayor parte de...   
6       P621_¿En qué actividad ocupa la mayor parte de...   
9       P621_¿En qué actividad ocupa la mayor parte de...   
10      P621_¿En qué actividad ocupa la mayor parte de...   
11      P621_¿En qué actividad ocupa la mayor parte de...   
...                                                   ...   
286589                             P6211_Cual_Normalizada   
286591                             P6211_Cual_Normalizada   
286602                             P6211_Cual_Normalizada   
286662                             P6211_Cual_Normalizada   
286685                             P6211_Cual_Normalizada   

                                                Respuesta

In [11]:
df_unificado.to_excel("dato_prueba.xlsx", index=False)

KeyboardInterrupt: 

In [ ]:
import pandas as pd
from functools import reduce

# 1. Columnas Identificadoras
cols_fijas = [
    'ID', 'START', 'END', 'MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta', 
    'LOCALIDAD', 'Sexo al nacer', 'Según factura de energía eléctrica estrato', 
    'Rango de edad', 'Edad exacta'
]

# 2. Definición de grupos (Asegúrate de que los nombres de las llaves sean distintos a las columnas originales)
grupos = {
    'Resultado_Actividad': [
        'P621_¿En qué actividad ocupa la mayor parte del tiempo?  ',
        'P622_Desempleado', 'P623_Estudiando', 'P624_Actividades de ocio y/o recreación',
        'P625_Oficios del Hogar', 'P626_Buscando trabajo', 
        'P627_Incapacitado Permanente para Trabajar', 'P628_Otro', 
        'P629_No sabe', 'P6210_No responde', 'P6211_Cual_Normalizada'
    ],
    'Resultado_Delito': [
        'P1171_De las siguientes modalidades delictivas, ¿Cuáles son las tres que más le preocupa?',
        'P1172_Hurto a residencias', 'P1173_Homicidio', 'P1174_Lesiones personales ', 
        'P1175_Hurto de vehículos (o alguna de sus partes)', 'P1176_Venta de drogas', 
        'P1177_Violencia intrafamiliar', 'P1178_Secuestro', 'P1179_Violencia sexual', 
        'P11711_Bloqueos (Obstrucción a vías públicas que afecten el orden público) ', 
        'P11712_Extorsión', 'P11712_Hurto a establecimientos comerciales', 
        'P11713_Vandalismo', 'P11714_Terrorismo', 'P11715_Trata de personas', 
        'P11716_Violencias contra la mujer (física, psicológica o económica)', 
        'P11717_Desaparición forzada', 'P11718_Desplazamiento forzado intraurbano', 
        'P11719_Amenazas', 'P11720_Feminicidio (asesinato de mujeres)', 'P11721_Abigeato'
    ],
    'Resultado_Fuente': [
        'P1281_Con relación a la seguridad de Bogotá ¿Cuáles diría usted que son las tres principales fuentes de información que le permite formar una opinión sobre los temas de seguridad en el municipio?  ',
        'P1282_Experiencias de familiares', 'P1283_Comentarios de la gente (voz a voz).', 
        'P1284_Redes sociales (Twitter, Facebook, grupos de Whatsapp).', 
        'P1285_Noticieros de Televisión.', 'P1286_Revistas (impresas o digitales).', 
        'P1287_Radio.', 'P1288_Periódicos (impresos o digitales).', 
        'P1289_Páginas oficiales en internet (Alcaldía, Fiscalía, Policía, Medicina Legal).'
    ]
}

dfs_procesados = []

for nombre_col, cols_grupo in grupos.items():
    # 3. Verificamos que las columnas existan en el DF antes de procesar
    cols_presentes = [c for c in cols_grupo if c in df_EPV.columns]
    
    if cols_presentes:
        # Transformamos a formato largo
        temp = pd.melt(df_EPV, id_vars=cols_fijas, value_vars=cols_presentes, value_name=nombre_col)
        
        # Eliminamos nulos para que solo queden las opciones marcadas
        temp = temp.dropna(subset=[nombre_col])
        
        # Seleccionamos columnas finales para este grupo
        dfs_procesados.append(temp[cols_fijas + [nombre_col]])

# 4. Unión de todos los dataframes (Merge Outer)
# Esto garantiza que si el ID 1 tiene 2 actividades y 1 delito, salgan 2 filas con ese delito repetido
df_EPV_final = reduce(lambda left, right: pd.merge(left, right, on=cols_fijas, how='outer'), dfs_procesados)

print("¡Proceso completado!")
print(f"Dimensiones finales: {df_EPV_final.shape}")
print(df_EPV_final.head())

In [36]:
df_EPV_final


ID                       START                         END  \
0           1         2024-06-04 14:40:25         2024-06-04 14:52:58   
1           1         2024-06-04 14:40:25         2024-06-04 14:52:58   
2           1         2024-06-04 14:40:25         2024-06-04 14:52:58   
3           1         2024-06-04 14:40:25         2024-06-04 14:52:58   
4           1         2024-06-04 14:40:25         2024-06-04 14:52:58   
...       ...                         ...                         ...   
200540  26716  13/07/2024  10:33:32 a. m.  13/07/2024  11:57:38 a. m.   
200541  26717   11/07/2024  2:48:35 p. m.   11/07/2024  3:00:59 p. m.   
200542  26717   11/07/2024  2:48:35 p. m.   11/07/2024  3:00:59 p. m.   
200543  26717   11/07/2024  2:48:35 p. m.   11/07/2024  3:00:59 p. m.   
200544  26717   11/07/2024  2:48:35 p. m.   11/07/2024  3:00:59 p. m.   

       MUNICIPIO_Seleccione el municipio donde va a aplicar la encuesta  \
0                                                  Bogotá                 
1                                                  Bogotá                 
2                                                  Bogotá                 
3                                                  Bogotá                 
4                                                  Bogotá                 
...                                                   ...                 
200540                                             Bogotá                 
200541                                             Bogotá                 
200542                                             Bogotá                 
200543                                             Bogotá                 
200544                                             Bogotá                 

          LOCALIDAD Sexo al nacer  Según factura de energía eléctrica estrato  \
0              Suba      Femenino                                           2   
1              Suba      Femenino                                           2   
2              Suba      Femenino                                           2   
3              Suba      Femenino                                           2   
4              Suba      Femenino                                           2   
...             ...           ...                                         ...   
200540  Teusaquillo      Femenino                                           3   
200541   Tunjuelito     Masculino                                           3   
200542   Tunjuelito     Masculino                                           3   
200543   Tunjuelito     Masculino                                           3   
200544   Tunjuelito     Masculino                                           3   

          Rango de edad  Edad exacta  \
0       De 25 a 35 años           29   
1       De 25 a 35 años           29   
2       De 25 a 35 años           29   
3       De 25 a 35 años           29   
4       De 25 a 35 años           29   
...                 ...          ...   
200540  De 46 a 55 años           52   
200541  De 36 a 45 años           44   
200542  De 36 a 45 años           44   
200543  De 36 a 45 años           44   
200544  De 36 a 45 años           44   

                                      Resultado_Actividad  \
0                                              Estudiando   
1                                              Estudiando   
2                                              Estudiando   
3                                              Estudiando   
4                                              Estudiando   
...                                                   ...   
200540  Trabajando (empleado, independiente, cuenta pr...   
200541  Trabajando (empleado, independiente, cuenta pr...   
200542  Trabajando (empleado, independiente, cuenta pr...   
200543  Trabajando (empleado, independiente, cuenta pr...   
200544  Trabajando (empleado, independiente, cuenta pr...   

                                         Resul

In [38]:
df_EPV_final["Resultado_Fuente"].value_counts()

Resultado_Fuente
Noticieros de Televisión.                                                       43650
Comentarios de la gente (voz a voz).                                            40078
Redes sociales (Twitter, Facebook, grupos de Whatsapp).                         39153
Experiencia personal                                                            30525
Experiencias de familiares                                                      21495
Radio.                                                                          13307
Páginas oficiales en internet (Alcaldía, Fiscalía, Policía, Medicina Legal).     5160
Periódicos (impresos o digitales).                                               5033
Revistas (impresas o digitales).                                                 2144
Name: count, dtype: int64